src/object/obj_tx.c

```diff
@@ -1179,11 +1175,179 @@ dc_tx_add_punch_akeys(struct dc_tx *tx, daos_handle_t oh, uint64_t flags,
+					  DAOS_COND_AKEY_UPDATE)) {
+			if (task->dt_result != 0)
+				D_GOTO(out, rc = task->dt_result);
+		}
+
+		rc = dc_tx_add_update(tx, args->oh, args->flags,
+				      args->dkey, args->nr,
+				      args->iods_or_akeys, args->sgls);
+		break;
+	case DAOS_OBJ_RPC_PUNCH_DKEYS:
+		D_ASSERT(args->flags & DAOS_COND_PUNCH);
+
+		if (task->dt_result != 0)
+			D_GOTO(out, rc = task->dt_result);
+
+		rc = dc_tx_add_punch_dkey(tx, args->oh, args->flags,
+					  args->dkey);
+		break;
+	case DAOS_OBJ_RPC_PUNCH_AKEYS:
+		D_ASSERT(args->flags & DAOS_COND_PUNCH);
+
+		if (task->dt_result != 0)
+			D_GOTO(out, rc = task->dt_result);
+
+		rc = dc_tx_add_punch_akeys(tx, args->oh, args->flags,
+					   args->dkey, args->nr,
+					   args->iods_or_akeys);
+		break;
+	default:
+		D_ASSERT(0);
+	}
+
+out:
+	D_MUTEX_UNLOCK(&tx->tx_lock);
+
+	/* Drop the reference that is held via dc_tx_attach(). */
+	dc_tx_decref(tx);
+	if (args->tmp_iods != NULL) {
+		int		i;
+
+		for (i = 0; i < args->nr; i++)
+			daos_iov_free(&args->tmp_iods[i].iod_name);
+
+		D_FREE(args->tmp_iods);
+	}
+
+	D_FREE(args);
+	/* The errno will be auto propagated to the dependent task. */
+	task->dt_result = rc;
+}
+
+static int
+dc_tx_check_existence_task(enum obj_rpc_opc opc, daos_handle_t oh,
+			   struct dc_tx *tx, uint64_t flags, daos_key_t *dkey,
+			   uint32_t nr, void *iods_or_akeys, d_sg_list_t *sgls,
+			   tse_task_t *parent)
+{
+	daos_iod_t				*iods = NULL;
+	struct dc_tx_check_existence_cb_args	*cb_args;
+	tse_task_t				*task;
+	uint64_t				 api_flags;
+	int					 rc;
+	int					 i;
+
+	D_ALLOC_PTR(cb_args);
+	if (cb_args == NULL)
+		return -DER_NOMEM;
+
+	cb_args->opc		= opc;
+	cb_args->tx		= tx;
+	cb_args->oh		= oh;
+	cb_args->flags		= flags;
+	cb_args->dkey		= dkey;
+	cb_args->nr		= nr;
+	cb_args->iods_or_akeys	= iods_or_akeys;
+	cb_args->sgls		= sgls;
+
+	if (nr != 0) {
+		D_ASSERT(iods_or_akeys != NULL);
+
+		if (opc != DAOS_OBJ_RPC_UPDATE) {
+			D_ALLOC_ARRAY(iods, nr);
+			if (iods == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			for (i = 0; i < nr; i++) {
+				rc = daos_iov_copy(&iods[i].iod_name,
+					&((daos_key_t *)iods_or_akeys)[i]);
+				if (rc != 0)
+					goto out;
+			}
+
+			api_flags = DAOS_COND_AKEY_FETCH;
+			cb_args->tmp_iods = iods;
+		} else if (flags & (DAOS_COND_AKEY_INSERT |
+				    DAOS_COND_AKEY_UPDATE)) {
+			iods = iods_or_akeys;
+			api_flags = DAOS_COND_AKEY_FETCH;
+		} else {
+			/* Only check dkey existence. */
+			api_flags = DAOS_COND_DKEY_FETCH;
+		}
+	} else {
+		api_flags = DAOS_COND_DKEY_FETCH;
+	}
+
+	rc = dc_obj_fetch_task_create(oh, dc_tx_ptr2hdl(tx), api_flags, dkey,
+			nr, DIOF_CHECK_EXISTENCE | DIOF_TO_LEADER, iods, sgls,
+			NULL, cb_args, NULL, tse_task2sched(parent), &task);
+	if (rc != 0)
+		goto out;
+
+	rc = dc_task_depend(parent, 1, &task);
+	if (rc != 0) {
+		D_ERROR("Fail to add dep on check existence task: "DF_RC"\n",
+			DP_RC(rc));
+		dc_task_decref(task);
+		goto out;
+	}
+
+	return dc_task_schedule(task, true);
```

Comment:
dc_obj_update_task and obj_punch_common call tse_task_complete after calling dc_tx_attach. Does it mean that we will complete the event and potentially reading users' buffers later when the existence check RPC is complete?

Reply:
The dc_tx_attach() caller for conditional modification should not call tse_task_complete(). I have pushed another patch to fix that.